taiga_halo2/examples/taiga_sudoku/app_vp.rs

```diff
@@ -0,0 +1,800 @@
+            &config.state_update_config,
+            &config.triple_mul_config,
+            &config.value_check_config,
+            &is_spend_note,
+            &previous_sudoku_cells,
+            &current_sudoku_cells,
+            &spend_note_variables[0],
+            &output_note_variables[0],
+        )?;
+
+        Ok(())
+    }
+}
+
+vp_circuit_impl!(SudokuAppValidityPredicateCircuit);
+
+#[test]
+fn test_halo2_sudoku_app_vp_circuit_init() {
+    use halo2_proofs::dev::MockProver;
+    use rand::rngs::OsRng;
+
+    let mut rng = OsRng;
+    let circuit = SudokuAppValidityPredicateCircuit::dummy(&mut rng);
+    let instances = circuit.get_instances();
+
+    let prover = MockProver::<pallas::Base>::run(13, &circuit, vec![instances]).unwrap();
+    assert_eq!(prover.verify(), Ok(()));
+}
+
+#[test]
+fn test_halo2_sudoku_app_vp_circuit_update() {
+    use halo2_proofs::dev::MockProver;
+    use rand::rngs::OsRng;
+
+    let mut rng = OsRng;
+    // Construct circuit
+    let circuit = {
+        let mut spend_notes = [(); NUM_NOTE].map(|_| Note::dummy(&mut rng));
+        let mut output_notes = [(); NUM_NOTE].map(|_| Note::dummy(&mut rng));
+        let is_spend_note = pallas::Base::one();
+        let init_state = SudokuState {
+            state: [
+                [5, 0, 1, 6, 7, 2, 4, 3, 9],
+                [7, 0, 2, 8, 4, 3, 6, 5, 1],
+                [3, 0, 4, 5, 9, 1, 7, 8, 2],
+                [4, 0, 8, 9, 5, 7, 2, 1, 6],
+                [2, 0, 6, 1, 8, 4, 9, 7, 3],
+                [1, 0, 9, 3, 2, 6, 8, 4, 5],
+                [8, 0, 5, 2, 1, 9, 3, 6, 7],
+                [9, 0, 3, 7, 6, 8, 5, 2, 4],
+                [6, 0, 7, 4, 3, 5, 1, 9, 8],
+            ],
+        };
+        let encoded_init_state = init_state.encode();
+        let previous_state = SudokuState {
+            state: [
+                [5, 8, 1, 6, 7, 2, 4, 3, 9],
+                [7, 9, 2, 8, 4, 3, 6, 5, 1],
+                [3, 0, 4, 5, 9, 1, 7, 8, 2],
+                [4, 0, 8, 9, 5, 7, 2, 1, 6],
+                [2, 0, 6, 1, 8, 4, 9, 7, 3],
+                [1, 0, 9, 3, 2, 6, 8, 4, 5],
+                [8, 0, 5, 2, 1, 9, 3, 6, 7],
+                [9, 0, 3, 7, 6, 8, 5, 2, 4],
+                [6, 0, 7, 4, 3, 5, 1, 9, 8],
+            ],
+        };
+        let current_state = SudokuState {
+            state: [
+                [5, 8, 1, 6, 7, 2, 4, 3, 9],
+                [7, 9, 2, 8, 4, 3, 6, 5, 1],
+                [3, 6, 4, 5, 9, 1, 7, 8, 2],
+                [4, 3, 8, 9, 5, 7, 2, 1, 6],
+                [2, 0, 6, 1, 8, 4, 9, 7, 3],
+                [1, 0, 9, 3, 2, 6, 8, 4, 5],
+                [8, 0, 5, 2, 1, 9, 3, 6, 7],
+                [9, 0, 3, 7, 6, 8, 5, 2, 4],
+                [6, 0, 7, 4, 3, 5, 1, 9, 8],
+            ],
+        };
+        spend_notes[0].value_base.app_data =
+            poseidon_hash(encoded_init_state, previous_state.encode());
+        spend_notes[0].value = 1u64;
```

Comment:
We seem to be constraining the value of the output note to 0 or 1. Are we constraining that the value of the spend note is 1? What if it is another value, like 6?

Reply:
I have the constraint.
https://github.com/anoma/taiga/blob/8d7c6c975e12128478b48431b5f468c970e126d2/taiga_halo2/examples/taiga_sudoku/gadgets/value_check.rs#L51

Reply:
The tests seem to pass if I set the value of the spend note to 6 while the value of the output note is 1

Reply:
Oh, I see.
I only constrained the output note.
It shouldn't be a problem if the value of any created output notes is constrained to 0 or 1. Because the spend notes are valid output notes created before. Do you think it makes sense? Or do we need to add the value constraint for the spend note?

Reply:
I need to think about it further. We don't check the existence of the spend note since its `is_merkle_checked` is false. 
There could be a problem here

Reply:
I think you're right. We'd better constrain the value of the spend note to be 1 (the note with 0 value will not be spent). The only risk I can come up with is that the spend note value can be set as arbitrary(like in your tests) if we don't have the constraint, but the total balance check will not pass.
I'll fix it tomorrow.